src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/failover/ClusterAwareMetricsContainer.java

```diff
@@ -0,0 +1,165 @@
+package com.mysql.cj.jdbc.ha.plugins.failover;
+
+import com.mysql.cj.conf.HostInfo;
+import com.mysql.cj.conf.PropertyKey;
+import com.mysql.cj.conf.PropertySet;
+import com.mysql.cj.jdbc.JdbcConnection;
+import com.mysql.cj.jdbc.ha.plugins.ICurrentConnectionProvider;
+import com.mysql.cj.log.Log;
+import com.mysql.cj.util.StringUtils;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class ClusterAwareMetricsContainer implements IClusterAwareMetricsContainer {
+    // ClusterID, Metrics
+    private final static Map<String, ClusterAwareMetrics> clusterMetrics = new ConcurrentHashMap<>();
+    // Instance URL, Metrics
+    private final static Map<String, ClusterAwareMetrics> instanceMetrics = new ConcurrentHashMap<>();
+
+    private ICurrentConnectionProvider currentConnectionProvider;
+    private PropertySet propertySet;
+    private String clusterID;
+
+    public ClusterAwareMetricsContainer(ICurrentConnectionProvider currentConnectionProvider, PropertySet propertySet) {
+        this.currentConnectionProvider = currentConnectionProvider;
+        this.propertySet = propertySet;
+    }
+
+    public static void linkInstances(List<HostInfo> hosts, String clusterID) {
+        final ClusterAwareMetrics metrics = clusterMetrics.computeIfAbsent(
+            clusterID,
+            k -> new ClusterAwareMetrics());
+        for (HostInfo host : hosts) {
+            clusterMetrics.put(host.getHost(), metrics);
+        }
+    }
+
+    public void setClusterID(String clusterID) {
+        this.clusterID = !StringUtils.isNullOrEmpty(clusterID) ? clusterID.substring(0, clusterID.indexOf(':'))
+            : getCurrentConnUrl();
+    }
+
+    public void registerFailureDetectionTime(long timeMs) {
+        if (!propertySet.getBooleanProperty(PropertyKey.gatherPerfMetrics.getKeyName()).getValue()) return;
+
+        getClusterMetrics(clusterID).registerFailureDetectionTime(timeMs);
+
+        if (propertySet.getBooleanProperty(PropertyKey.gatherAdditionalMetricsOnInstance.getKeyName()).getValue()) {
+            getInstanceMetrics(getCurrentConnUrl()).registerFailureDetectionTime(timeMs);
+        }
+    }
+
+    public void registerWriterFailoverProcedureTime(long timeMs) {
+        if (!propertySet.getBooleanProperty(PropertyKey.gatherPerfMetrics.getKeyName()).getValue()) return;
```

Comment:
Please always wrap your if statements in clauses, and no single line if statements please.